v2/index_gen.go

```diff
@@ -28,21 +31,63 @@ func GenerateIndex(v1r io.Reader, opts ...Option) (index.Index, error) {
 	return idx, nil
 }
 
-// LoadIndex populates idx with index records generated from v1r.
-// The v1r must be data payload in CARv1 format.
-func LoadIndex(idx index.Index, v1r io.Reader, opts ...Option) error {
-	reader := internalio.ToByteReadSeeker(v1r)
-	header, err := carv1.ReadHeader(reader)
+// LoadIndex populates idx with index records generated from r.
+// The r may be in CARv1 or CARv2 format.
+//
+// Note, the index is re-generated every time even if r is in CARv2 format and already has an index.
+// To read existing index when available see ReadOrGenerateIndex.
+func LoadIndex(idx index.Index, r io.Reader, opts ...Option) error {
+	reader := internalio.ToByteReadSeeker(r)
+	pragma, err := carv1.ReadHeader(r)
 	if err != nil {
 		return fmt.Errorf("error reading car header: %w", err)
 	}
 
-	if header.Version != 1 {
-		return fmt.Errorf("expected version to be 1, got %v", header.Version)
-	}
+	var dataSize, dataOffset int64
+	switch pragma.Version {
+	case 1:
+		break
+	case 2:
+		// Read V2 header which should appear immediately after pragma according to CARv2 spec.
+		var v2h Header
+		_, err := v2h.ReadFrom(r)
+		if err != nil {
+			return err
+		}
 
-	// Parse Options.
-	o := ApplyOptions(opts...)
+		// Sanity-check the CARv2 header
+		if v2h.DataOffset < HeaderSize {
+			return fmt.Errorf("malformed CARv2; data offset too small: %d", v2h.DataOffset)
+		}
+		if v2h.DataSize < 1 {
```

Comment:
I think a zero-length payload is still valid isn't it? it would just produce an empty index. not a big deal, zero payload probably means an error but strictly speaking I think `0` should be acceptable.

Reply:
I thought about this when writing this statement. I actually think `1` might be too small. The rationale is that according to CARv2 spec the data size refers to the inner CARv1; as in the entire thing including the CARv1 header. At the danger of sounding pedantic (apologies if I do 🙂) accepting zero as `DataSize` would then mean an empty file is technically a valid CARv1? 

That's why I think that value should be the minimum possible size for a CARv1, which is a CARv1 header with version value 1 and no roots.

WDYT?

Reply:
Looks like the CARv1 should have at [least one root](https://github.com/ipld/go-car/blob/master/car.go#L138) CID, which increases the minimum acceptable CARv1 size.